src/antimeridian/_implementation.py

```diff
@@ -212,39 +212,55 @@ def build_polygons(
     if not segments:
         return []
     segment = segments.pop()
-    right = (
-        segment[-1][0] == 180
-    )  # all segments should start and end at abs(180) longitude
-    candidates: List[
-        Tuple[Optional[int], float, bool]
-    ] = list()  # list of (index, latitude, is_start)
-    if segment[0][0] == segment[-1][0] and (
-        (right and segment[0][1] > segment[-1][1])
-        or (not right and segment[0][1] < segment[-1][1])
-    ):
-        candidates.append((None, segment[0][1], True))
+    is_right = segment[-1][0] == 180
+    candidates: List[Tuple[Optional[int], float]] = list()
+    if is_self_closing(segment):
+        candidates.append((None, segment[0][1]))
     for i, s in enumerate(segments):
         if s[0][0] == segment[-1][0]:
-            if (right and s[0][1] > segment[-1][1]) or (
-                not right and s[0][1] < segment[-1][1]
+            # The logic here is:
+            # - If s is on our side, and
+            # - s's start is closer to the pole than we are, and
+            # - s is not-self closing, or
+            # - s is self closing but its start is closer to us than its end, or
+            # - its end is further away from the pole than our start
+            if (
+                is_right
+                and s[0][1] > segment[-1][1]
+                and (
+                    not is_self_closing(s)
+                    or s[0][1] < s[-1][1]
+                    or segment[-1][1] < s[0][1]
```

Comment:
This is a great illustration of why I miss [zero-cost abstractions](https://doc.rust-lang.org/beta/embedded-book/static-guarantees/zero-cost-abstractions.html). I don't want to convert every point to a `Point` structure, since I know that conversion has performance penalties, but the non-structured data makes this algorithm really hard to read. 🤷🏼 🦀